src/LtiAssignmentsGradesService.php

```diff
@@ -52,7 +53,10 @@ public function findOrCreateLineitem(LtiLineitem $newLineItem)
                 (isset($lineitem['resourceId']) && $lineitem['resourceId'] == $newLineItem->getResourceId()) ||
                 (isset($lineitem['resourceLinkId']) && $lineitem['resourceLinkId'] == $newLineItem->getResourceLinkId())
             ) {
-                if (empty($newLineItem->getTag()) || $lineitem['tag'] == $newLineItem->getTag()) {
+                if (
+                    empty($newLineItem->getTag()) ||
+                    (isset($lineitem['tag']) && $lineitem['tag'] == $newLineItem->getTag())
+                ) {
```

Comment:
This whole nested `if` block is a bit if a mess. Let's simplify it. Here, let's just do:
```php
        foreach ($lineitems as $lineitem) {
            if ($this->isMatchingLineitem($lineitem, $newLineItem)) {
                return new LtiLineitem($lineitem);
            }
        }
```

Then, down below add a private method:
```php
    private function isMatchingLineitem(array $lineitem, LtiLineitem $newLineItem): bool
    {
        return (
                (empty($newLineItem->getResourceId()) && empty($newLineItem->getResourceLinkId())) ||
                (isset($lineitem['resourceId']) && $lineitem['resourceId'] == $newLineItem->getResourceId()) ||
                (isset($lineitem['resourceLinkId']) && $lineitem['resourceLinkId'] == $newLineItem->getResourceLinkId())
            ) && (
                empty($newLineItem->getTag()) || \(isset($lineitem['tag']) && $lineitem['tag'] == $newLineItem->getTag())
            );
    }
```

Reply:
Sounds good -- honestly, I just kept the formatting that was there. I am not even sure these conditions are correct -- do you know how this code was arrived at? It seems like if the new line item has no resource id, no resource link id, and no tag, it will select the first line item regardless of whether that line item has those fields or not. Is this accurate? It seems like if the new line item is missing those fields it should require that the returned line item also has those fields missing, but I am not an expert in LTI 1.3 line items -- it just feels off and it didn't seem like other libraries implemented the checks in this way.

Reply:
The logic may well be wrong. If you're willing to help me untangle it, I'll try and write some tests for it. Does this look right?

`$this->isMatchingLineitem()` should return `true` if:

- The resource ID and tag match on both lineitems
- The resource link ID and tag match on both lineitems

`$this->isMatchingLineitem()` should return `false` if:

- The resource ID don't match
- The resource link ID don't match
- The resource ID matches but the tag doesn't
- The resource link ID matches but the tag doesn't
- The `$newLineItem` is an empty object (i.e. doesn't have a resource ID or resource link ID)

If I'm understanding you correctly, that last condition in the list currently returns `true`, when it should be `false`. If that's the case, I think we can just remove all of the `empty($blah)` statements.

> do you know how this code was arrived at? ... Is this accurate? ... it just feels off and it didn't seem like other libraries implemented the checks in this way.

I inherited this project. I didn't write this specific logic, nor am I confident in its accuracy. We're using some of this code in our platform for gradebook sync, but in our code we always set the lineitem via the serviceData. I don't think our code actually uses this part, so you're probably the first person to actually put it to real use. It's entirely possible this logic is wrong. If it's helpful, here's how we're handling it in our code:

```php
        if (!isset($endpointData['lineitem'])) {
            $scoreLineItem = LtiLineitem::new()
                ->setTag('score')
                ->setScoreMaximum(100)
                ->setLabel('Some assignment title')
                ->setResourceLinkId($resourceLink->id);

            $lineItem = $gradeService->findOrCreateLineitem($scoreLineItem);
            $result->lineitem_url = $lineItem->getId();
            $result->save();

            $endpointData['lineitem'] = $result->lineitem_url;
            // reinitialize grades to pick up new endpointData
            $gradeService->setServiceData($endpointData);
        }

        $score = LtiGrade::new()
                ->setScoreGiven($grade->score)
                ->setScoreMaximum(100)
                ->setTimestamp(Carbon::now()->toIso8601String())
                ->setActivityProgress('Completed')
                ->setGradingProgress('FullyGraded')
                ->setUserId($user->lti_user_id)
                ->setCanvasExtension(['submitted_at' => $grade->submittedAt]);

        $response = $gradeService->putGrade($score);
```

Reply:
Interesting, thanks for showing how you’re using it! I think that I will also only be using the service data to post back to the default line item. I was merely testing out how things worked with creating custom line items when I realized the “get all line items” call was not working which would also break the “find or create line item” as it was just creating a new line item every time.

I’m not sure I know enough about the ways various LTI platforms respond to LTI tools to know the best way to do isMatchingLineitem. If it was how I would expect it, and based on canvas docs, the line items would only match if tag, resource id and resource link id are each either empty in both line items or match values in both line items.

I’m on my phone now, but could write up some sample code in a day or two.

Reply:
My instinct would be code like the following (untested):

```php
private function isMatchingLineitem(array $lineitem, LtiLineitem $newLineItem): bool
{
    if (($lineitem['tag'] ?? null) != $newLineItem->getTag()) { return false; }
    if (($lineitem['resourceId'] ?? null) != $newLineItem->getResourceId()) { return false; }
    if (($lineitem['resourceLinkId'] ?? null) != $newLineItem->getResourceLinkId()) { return false; }
    return true;
}
```

If we're trying to match a line-item, I'm not sure why you wouldn't want the tool-defined fields (tag and resourceId) to both match up or not be defined, and the resourceLinkId (which is like an assignment in the LMS) to also match or be undefined. Thoughts? Again, I am no expert in LTI integrations, just thinking through the scenarios where I would use multiple line items and I think you'd want a match on all three params.

Reply:
This looks sensible to me, and I agree with you on the logic.